crates/ra_assists/src/handlers/replace_unwrap_with_match.rs

```diff
@@ -0,0 +1,195 @@
+use std::iter;
+
+use hir::name;
+use ra_syntax::{
+    ast::{self, make},
+    AstNode,
+};
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::edit::IndentLevel;
+
+// Assist: replace_unwrap_with_match
+//
+// Replaces `unwrap` a `match` expression. Works for Result and Option.
+//
+// ```
+// enum Result<T, E> { Ok(T), Err(E) }
+// fn main() {
+//     let x: Result<i32, i32> = Result::Ok(92);
+//     let y = x.<|>unwrap();
+// }
+// ```
+// ->
+// ```
+// enum Result<T, E> { Ok(T), Err(E) }
+// fn main() {
+//     let x: Result<i32, i32> = Result::Ok(92);
+//     let y = match x {
+//         Ok(a) => a,
+//         _ => unreachable!(),
+//     };
+// }
+// ```
+pub(crate) fn replace_unwrap_with_match(ctx: AssistCtx) -> Option<Assist> {
+    let method_call: ast::MethodCallExpr = ctx.find_node_at_offset()?;
+    let name = method_call.name_ref()?;
+    if name.text() != "unwrap" {
+        return None;
+    }
+    let caller = method_call.expr()?;
+    let ty = ctx.sema.type_of_expr(&caller)?;
+
+    let type_name = ty.as_adt()?.name(ctx.sema.db);
+    if type_name == name![Result] {
```

Comment:
Rather then using `name!`, let's do `type_name.to_string() == "Result"`. `name!` should be mostly internal. In particular, we probably shouldn't add `Option` there just for the sake of a single assit.

Reply:
We could actually make this completely agnostic of the particular enum by looking at the code of the `unwrap` function and using the branches from there, assuming it's a single `match` (which it is for `Option` and `Result`)... I'm not suggesting we do this right now though :sweat_smile:

Reply:
Hmm, yeah, you know going further an assist to **inline** the function call should actually be a superset of this assist!

Reply:
Well, except that at least the `Result` implementation calls a private function instead of directly `panic`ing :sweat_smile:

Reply:
bruh